### ISSUE TITLE
Macro for constructing stateful cells?

### ISSUE THREAD

**Issue by jackfirth** [zyrolasting/kinda-ferpy#2]
The `stateful-cell` function seems to lead to a lot of indentation when used with nontrivial computations:

```racket
(define positive-or-negative
  (stateful-cell
   (λ _
     (let ([p (positive)]
           [n (negative)])
       (if (switch) p n)))))
```

What do you think of providing a macro that accepted a series of body expressions? So that the above could be written like this:

```racket
(define positive-or-negative
  (stateful-cell
    (define p (positive))
    (define n (negative))
    (if (switch) p n)))
```

Maybe with some sort of keyword clause for listing explicit dependencies?

```racket
(define positive-or-negative
  (stateful-cell
    #:dependency p positive
    #:dependency n negative
    (if (switch) p n)))
```

**Comment by zyrolasting** [comment]
I'd need to use a different name for stability reasons, but I am totally good with that. I'll have that ready as soon as possible.

**Comment by jackfirth** [comment]
Usually when I want a convenient macro for constructing a `foo` that
doesn't support every use case, and I want a regular function for
constructing a `foo` that does, I name the macro version `foo` and the
function version `make-foo`. Would that work?

I realize that means you would have to rename the current `stateful-cell`
to `make-stateful-cell` and break backwards compatibility, but
realistically the best time to make that kind of change is just after
publishing the initial release. In my experience, APIs that _never_ break
backwards compatibility simply become unusable and never improve.

**Comment by zyrolasting** [comment]
Fair enough. We're on version `0.x` anyway, so I'm good with the names. I'll get to this after #1

**Comment by zyrolasting** [comment]
Since the macro system is still new to me I ended up with a behemoth:

```
(define-syntax (stateful-cell stx)
  (syntax-case stx ()
    [(_ code ...)
     (let-values ([(parsed-options unused)
                   (parse-keyword-options #'(code ...)
                                          (list (list '#:dependency
                                                      check-identifier
                                                      check-identifier))
                                          #:context #'stateful-cell)])
       (with-syntax ([(requested-dependency-bindings ...)
                      (map (λ (spec)
                             (with-syntax ([orig/id (third spec)]
                                           [body/id (fourth spec)])
                               #'(body/id (orig/id))))
                           parsed-options)]
                     [(body ...) unused])
         #'(make-stateful-cell
            (λ ()
              (let (requested-dependency-bindings ...)
                body ...)))))]))
```

I'm good with cleaning this up, but I want to first be sure we're all ok with the syntax as-is.

Question: I'm considering a form like `(stateful-cell switch positive negative : (if (switch) (positive) (negative)))`, where `:` (or otherwise) separates cells from names. Is there something that would make this form problematic and `#:dependency p positive`, etc. worth typing each time? I'm okay with supporting `#:dependency`, but I can tell you that I'd probably write yet another macro on top of it before too long.

**Comment by zyrolasting** [comment]
Macro simplified after convo in Racket Slack. See #4 for merged implementation

**Comment by jackfirth** [comment]
> Question: I'm considering a form like `(stateful-cell switch positive negative : (if (switch) (positive) (negative)))`, where `:` (or otherwise) separates cells from names. Is there something that would make this form problematic and `#:dependency p positive`, etc. worth typing each time? I'm okay with supporting `#:dependency`, but I can tell you that I'd probably write yet another macro on top of it before too long. If I really wanted a new name for a cell value I'd probably just write `(define p (positive))` in the cell body.

I think that makes sense. The `(define p (positive))` approach seems better to me too, it's more consistent with regular racket code. What if `#:dependency` just took a single identifier?

```racket
(stateful-cell
    #:dependency positive
    #:dependency negative
    (if (switch) (positive) (negative))))
```

It could also bind it as a variable so that the body of the cell wouldn't need to surround the cell with parens:

```racket
(stateful-cell
    #:dependency positive
    #:dependency negative
    (if (switch) positive negative)))
```

**Comment by zyrolasting** [comment]
I like the first of your two code blocks better. The identifiers point to cells in both the head pattern and the body. If I made it so that `#:dependency` had to be an identifier for a stateful cell, but the bodies were for the values, I'd worry about conveying a false equivalency that would be harder to detect in larger programs.

There's also a slight re-write burden. Given `(define c (% (a) (b) (c)))`, adding one `#:dependency` for `a`, `b`, and `c` entails stripping parens from each instance in the body.

TL;DR: I think one identifier works, but it's meaning of the identifier should not change in the body.

**Comment by zyrolasting** [comment]
I implemented the above change in a PR. Since it's another breaking change I'll leave it up for a bit just in case there's any follow-up.
